src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs

```diff
@@ -258,24 +239,266 @@ private IISAwaitable DoFlushAsync()
+        // 2. Done reading and await all writes.
+        public async Task ConsumeAsync()
+        {
+            await ReadAndWriteLoopAsync();
+
+            // If we are done writing, complete the output pipe and return
+            // Input Pipe will be closed when ReadAndWriteLoopAsync returns
+            if (_doneWriting)
+            {
+                Output.Reader.Complete();
+                return;
+            }
+
+            await WriteLoopAsync();
+        }
+
+        public async Task ReadAndWriteLoopAsync()
+        {
+            // First we check if there is anything to write from the Output pipe
+            // If there is, we call WriteToIISAsync
+            while (!_doneReading)
+            {
+                // Check if Output pipe has anything to write to IIS.
+                if (Output.Reader.TryRead(out var readResult))
+                {
+                    var buffer = readResult.Buffer;
+                    var consumed = buffer.End;
+
+                    try
+                    {
+                        // If the output pipe is termniated, cancel the request.
+                        if (readResult.IsCancelled)
+                        {
+                            _doneWriting = true;
+                            break;
+                        }
+
+                        if (!buffer.IsEmpty)
+                        {
+                            // Write to IIS buffers
+                            // Guaranteed to write the entire buffer to IIS
+                            await WriteToIISAsync(buffer);
+                        }
+                        else if (readResult.IsCompleted)
+                        {
+                            break;
+                        }
+                        else
+                        {
+                            // Flush of zero bytes
+                            await FlushToIISAsync();
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        throw ex;
+                    }
+                    finally
+                    {
+                        // Always Advance the data pointer to the end of the buffer.
+                        Output.Reader.AdvanceTo(buffer.End);
+                    }
+                }
+
+               
+                // Check if there was an upgrade. If there is, we will replace the request and response bodies with
+                // two seperate loops
+                await CheckForUpgrade();
+
+                // Now we handle the read. 
+                var wb = Input.Writer.GetMemory();
+                _inputHandle = wb.Retain(true);
+
+                try
+                {
+                    // Lock around invoking ReadFromIISAsync as we don't want to call CancelIo
+                    // when calling read
+                    var read = await ReadFromIISAsync(wb.Length);
```

Comment:
Don't errors thrown from here leak? Does it ever make it to the caller of ReadAsync?

Reply:
Yeah they would leak. How should I propagate them to the user? The failure here will not sync up with user code.